Core/MIPS/IR/IRCompVFPU.cpp

```diff
@@ -0,0 +1,388 @@
+			int negate = (prefix >> (16 + i)) & 1;
+			int constants = (prefix >> (12 + i)) & 1;
+
+			// Unchanged, hurray.
+			if (!constants && regnum == i && !abs && !negate)
+				continue;
+
+			/*
+			// This puts the value into a temp reg, so we won't write the modified value back.
+			vregs[i] = fpr.GetTempV();
+			if (!constants) {
+				fpr.MapDirtyInV(vregs[i], origV[regnum]);
+				fpr.SpillLockV(vregs[i]);
+
+				// Prefix may say "z, z, z, z" but if this is a pair, we force to x.
+				// TODO: But some ops seem to use const 0 instead?
+				if (regnum >= n) {
+					WARN_LOG(CPU, "JIT: Invalid VFPU swizzle: %08x : %d / %d at PC = %08x (%s)", prefix, regnum, n, GetCompilerPC(), MIPSDisasmAt(GetCompilerPC()));
+					regnum = 0;
+				}
+
+				if (abs) {
+					fp.FABS(fpr.V(vregs[i]), fpr.V(origV[regnum]));
+					if (negate)
+						fp.FNEG(fpr.V(vregs[i]), fpr.V(vregs[i]));
+				} else {
+					if (negate)
+						fp.FNEG(fpr.V(vregs[i]), fpr.V(origV[regnum]));
+					else
+						fp.FMOV(fpr.V(vregs[i]), fpr.V(origV[regnum]));
+				}
+			} else {
+				fpr.MapRegV(vregs[i], MAP_DIRTY | MAP_NOINIT);
+				fpr.SpillLockV(vregs[i]);
+				fp.MOVI2F(fpr.V(vregs[i]), constantArray[regnum + (abs << 2)], SCRATCH1, (bool)negate);
+			}
+			*/
+		}
+	}
+
+	void IRJit::GetVectorRegsPrefixD(u8 *regs, VectorSize sz, int vectorReg) {
+		_assert_(js.prefixDFlag & JitState::PREFIX_KNOWN);
+
+		GetVectorRegs(regs, sz, vectorReg);
+		if (js.prefixD == 0)
+			return;
+
+		int n = GetNumVectorElements(sz);
+		for (int i = 0; i < n; i++) {
+			// Hopefully this is rare, we'll just write it into a reg we drop.
+			//if (js.VfpuWriteMask(i))
+			//	regs[i] = fpr.GetTempV();
+		}
+	}
+
+	void IRJit::ApplyPrefixD(const u8 *vregs, VectorSize sz) {
+		_assert_(js.prefixDFlag & JitState::PREFIX_KNOWN);
+		if (!js.prefixD)
+			return;
+
+		/*
+		int n = GetNumVectorElements(sz);
+		for (int i = 0; i < n; i++) {
+			if (js.VfpuWriteMask(i))
+				continue;
+
+			int sat = (js.prefixD >> (i * 2)) & 3;
+			if (sat == 1) {
+				// clamped = x < 0 ? (x > 1 ? 1 : x) : x [0, 1]
+				fpr.MapRegV(vregs[i], MAP_DIRTY);
+
+				fp.MOVI2F(S0, 0.0f, SCRATCH1);
+				fp.MOVI2F(S1, 1.0f, SCRATCH1);
+				fp.FMIN(fpr.V(vregs[i]), fpr.V(vregs[i]), S1);
+				fp.FMAX(fpr.V(vregs[i]), fpr.V(vregs[i]), S0);
+			} else if (sat == 3) {
+				// clamped = x < -1 ? (x > 1 ? 1 : x) : x [-1, 1]
+				fpr.MapRegV(vregs[i], MAP_DIRTY);
+
+				fp.MOVI2F(S0, -1.0f, SCRATCH1);
+				fp.MOVI2F(S1, 1.0f, SCRATCH1);
+				fp.FMIN(fpr.V(vregs[i]), fpr.V(vregs[i]), S1);
+				fp.FMAX(fpr.V(vregs[i]), fpr.V(vregs[i]), S0);
+			}
+		}
+		*/
+	}
+
+	void IRJit::Comp_SV(MIPSOpcode op) {
+		DISABLE;
+	}
+
+	void IRJit::Comp_SVQ(MIPSOpcode op) {
+		int imm = (signed short)(op & 0xFFFC);
+		int vt = (((op >> 16) & 0x1f)) | ((op & 1) << 5);
+		MIPSGPReg rs = _RS;
+
+		u8 vregs[4];
+		GetVectorRegs(vregs, V_Quad, vt);
+
+		switch (op >> 26) {
+		case 54: //lv.q
+		{
+			// TODO: Add vector load/store instruction to the IR
```

Comment:
Yeah, this will be the trickiest part of the whole thing. I can see a few options:

  * Take it all down to scalar, optimize, and then try to get back to SIMD in the backends that consume the IR
  * Implement already in the IR pass a similar approach that we have in the x64 backend, where we join things into SIMD ops and write back again, emitting friendly modern 4x1 SIMD ops but may be pretty far from optimal in places. That makes the job easy for the backends that consume the IR as they'll just need to map IR-simd to their own SIMD and regalloc, but may lose some opportunities.
* Emit direct analogues to the PSP's operations (with a few simplifications like eliminating the matrix ops) and let the backends do the same work they do now.

Really not sure what is best.